zio-http/shared/src/main/scala/zio/http/codec/HttpContentCodec.scala

```diff
@@ -0,0 +1,195 @@
+package zio.http.codec
+
+import scala.collection.immutable.ListMap
+
+import zio._
+
+import zio.stream.ZPipeline
+
+import zio.schema.Schema
+import zio.schema.codec._
+
+import zio.http._
+import zio.http.codec.HttpContentCodec.json.fromSchema
+
+final case class HttpContentCodec[A](
+  choices: ListMap[MediaType, BinaryCodec[A]],
+  schema: Schema[A],
+) { self =>
+
+  /**
+   * A right biased merge of two HttpContentCodecs.
+   */
+  def ++(that: HttpContentCodec[A]): HttpContentCodec[A] =
+    HttpContentCodec(choices ++ that.choices, schema)
+
+  def decodeRequest(request: Request): Task[A]    = {
+    val contentType =
+      request.header(Header.ContentType).getOrElse(Header.ContentType(MediaType.application.`json`))
+    lookup(contentType) match {
+      case Some(codec) =>
+        request.body.asChunk.flatMap { bytes =>
+          ZIO.fromEither(codec.decode(bytes))
+        }
+      case None        =>
+        ZIO.fail(throw new IllegalArgumentException(s"No codec found for content type $contentType"))
+    }
+  }
+  def decodeResponse(response: Response): Task[A] = {
+    val contentType =
+      response.header(Header.ContentType).getOrElse(Header.ContentType(MediaType.application.`json`))
+    lookup(contentType) match {
+      case Some(codec) =>
+        response.body.asChunk.flatMap { bytes =>
+          ZIO.fromEither(codec.decode(bytes))
+        }
+      case None        =>
+        ZIO.fail(throw new IllegalArgumentException(s"No codec found for content type $contentType"))
+    }
+  }
+
+  def encode(value: A): Either[String, Body] =
+    choices.headOption match {
+      case Some((mediaType, codec)) =>
+        Right(Body.fromChunk(codec.encode(value)).contentType(mediaType))
+      case None                     =>
+        Left("No codec found")
+    }
+
+  def only(mediaType: MediaType): HttpContentCodec[A] =
+    HttpContentCodec(
+      ListMap(
+        mediaType -> lookup(mediaType)
+          .getOrElse(
+            throw new IllegalArgumentException(s"MediaType $mediaType is not supported by $self"),
+          ),
+      ),
+      schema,
+    )
+
+  def only(mediaType: Option[MediaType]): HttpContentCodec[A] = {
+    mediaType match {
+      case Some(mediaType) =>
+        HttpContentCodec(
+          ListMap(
+            mediaType -> lookup(mediaType)
+              .getOrElse(
+                throw new IllegalArgumentException(s"MediaType $mediaType is not supported by $self"),
+              ),
+          ),
+          schema,
+        )
+      case None            =>
+        self
+    }
+  }
+
+  def onlyOrNone(mediaType: MediaType): Option[HttpContentCodec[A]] =
```

Comment:
This is awkward because `HttpContentCodec` already has a `None`, namely, when the `ListMap` is empty. It may be a better idea to simply have `isEmpty` and `nonEmpty` methods on `HttpContentCodec`.